### PR TITLE
regexp func: Allow regexp literal as first argument

### LINKED ISSUE #4148: Make regexps compile to a string outside the search context
As of the filing of this issue, Zed is currently at commit e9624c3.

With the arrival of the `regexp()` function via #2093, it has been observed that in the current implementation, the backslashes that often appear in regexps need to be escaped to be passed into the function as part of a string representation of the regexp. E.g.:

```
$ zq -version
Version: v1.2.0-85-ge9624c3b

$ echo '{"query": "www.brimdata.io"}' | zq -z 'yield regexp("(\\w+\\.\\w+)$", query)' -
["brimdata.io","brimdata.io"]
```

It has been proposed that we improve on this by having regular expressions compile to a string outside the search context. For instance in when writing a regex for the new regexp function it would be able to do:

```
regexp(/(\w+\.\w+)/", query)
```

### PR DESCRIPTION
**Pull request by mattnibs**
Allow the use of a regexp literal as the first argument for the regexp function.

Close #4148

### CODE REVIEW

File: compiler/parser/parser.peg [hunk 1]
```diff
@@ -721,6 +721,7 @@ Cast
 
 Function
   = Grep
+  / RegexpFunc
```
Comment (nwt) [inline compiler/parser/parser.peg:724]:
```suggestion
  // Special case to handle `regexp(/x/, y)`.
  / "regexp" __ "(" __ arg0Text:RegexpPattern __ "," __ arg1:Expr __ ")" where:WhereClause? {
      VAR(arg0) = MAP("kind": "Primitive", "type": "string", "text": arg0Text)
      RETURN(MAP("kind": "Call", "name": "regexp", "args": ARRAY(arg0, arg1), "where": where))
    }
```

File: docs/language/functions/regexp.md [hunk 2]
```diff
@@ -18,7 +18,7 @@ no match.
 
 Regexp returns an array of the match and its subexpressions:
 ```mdtest-command
-echo '"seafood fool friend"' | zq -z 'yield regexp("foo(.?) (\\w+) fr.*", this)' -
+echo '"seafood fool friend"' | zq -z 'yield regexp(/foo(.?) (\w+) fr.*/, this)' -
```
Comment (nwt) [inline docs/language/functions/regexp.md:21]: We should update the synopsis and description, too.

File: compiler/parser/parser.peg (hunk 1 above)
Comment (mattnibs) [reply compiler/parser/parser.peg:724]: Oh yeah good call 👍🏻

File: docs/language/functions/regexp.md [hunk 3]
```diff
@@ -5,20 +5,20 @@
...
+regexp(re: string|regexp, s: string) -> any
 ```
 ### Description
 The _regexp_ function returns an array of strings holding the text
-of the left most match of the regular expression string `re` and the
-matches of each parenthesized subexpression (also known as capturing groups)
-if there are any. A null value indicates
-no match.
+of the left most match of the regular expression `re`, which can be either
+a string value or [regular expression](https://zed.brimdata.io/docs/next/language/overview#711-regular-expressions),
```
Comment (nwt) [inline docs/language/functions/regexp.md:13]: The link in glob.md looks like this.
```suggestion
a string value or a [regular expression](../overview.md#711-regular-expressions),
```

**Review by nwt — APPROVED** [review]

File: docs/language/functions/regexp.md (hunk 3 above)
Comment (mattnibs) [reply docs/language/functions/regexp.md:13]: Oh, shoot that's right.

**Pull request merged**